src/librustc/middle/resolve_lifetime.rs

```diff
@@ -447,6 +447,14 @@ fn krate<'tcx>(tcx: TyCtxt<'_, 'tcx, 'tcx>) -> NamedRegionMap {
     map
 }
 
+fn sub_items_have_self_param(node: &hir::ItemKind) -> bool {
```

Comment:
Nit: can we get a comment here to make it more obvious what is happening?

Something like:

---

In traits, there is an implicit `Self` type parameter which comes before the generics. We have to account for this when computing the index of the other generic parameters. This function returns true if there is such an implicit parameter defined on the given item.

Reply:
Sounds good to me.